Move call for evidence pages from government-frontend

deborahchua opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Related PRs:
- https://github.com/alphagov/whitehall/pull/10329
- https://github.com/alphagov/government-frontend/pull/3725

## What

Copy the rendering code for call for evidence pages from government-frontend.

## Why

[Trello card](https://trello.com/c/GzNmv0Ag/647-move-callforevidence-from-government-frontend-to-frontend), [Jira issue PNP-6340](https://gov-uk.atlassian.net/browse/PNP-6340)

## Examples

## Welsh only call for evidence:
- https://www.gov.uk/government/calls-for-evidence/cerbydau-awtomataidd-datganiad-o-egwyddorion-diogelwch.cy

### "Open" call for evidence:
- https://www.gov.uk/government/calls-for-evidence/review-of-cqc-regulation-9a-visiting-and-accompanying-in-care-homes-hospitals-and-hospices
- https://www.gov.uk/government/calls-for-evidence/automated-vehicles-statement-of-safety-principles

### "Closed" call for evidence:
- https://www.gov.uk/government/calls-for-evidence/data-brokers-and-national-security
- https://www.gov.uk/government/calls-for-evidence/credit-union-common-bond-reform

### Call for evidence "outcome"
- https://www.gov.uk/government/calls-for-evidence/safety-of-journalists-call-for-evidence
- https://www.gov.uk/government/calls-for-evidence/review-of-civil-legal-aid-call-for-evidence

## Screenshots?

### Welsh only call for evidence
https://www.gov.uk/government/calls-for-evidence/cerbydau-awtomataidd-datganiad-o-egwyddorion-diogelwch.cy

![Screenshot 2025-06-17 at 12 26 31](https://github.com/user-attachments/assets/bc1a40e9-0a13-4d28-beea-8c04c39319e8)
![Screenshot 2025-06-17 at 13 39 39](https://github.com/user-attachments/assets/95097663-7a9e-45a2-93a5-1071e11ea86c)

### "Open" Call for evidence
https://www.gov.uk/government/calls-for-evidence/review-of-cqc-regulation-9a-visiting-and-accompanying-in-care-homes-hospitals-and-hospices

![open_call_for_evidence-1](https://github.com/user-attachments/assets/d724965d-812a-41f0-8991-b3a127083ea1)
![open_call_for_evidence-2](https://github.com/user-attachments/assets/2ee43e06-f078-488d-a89e-68c5f06c5186)

### "Closed" Call for evidence
https://www.gov.uk/government/calls-for-evidence/data-brokers-and-national-security

![closed_call_for_evidence-1](https://github.com/user-attachments/assets/faceb82c-e952-45ce-980c-0f0c1b04e600)
![closed_call_for_evidence-2](https://github.com/user-attachments/assets/f0a5b1e1-1121-4f7d-9fd2-ffb837c95410)

### Call for evidence "outcome"
https://www.gov.uk/government/calls-for-evidence/youth-vaping-call-for-evidence

![call_for_evidence_outcome-1](https://github.com/user-attachments/assets/a68d7d35-9245-4fa4-a0de-a975287c8386)
![call_for_evidence_outcome-2](https://github.com/user-attachments/assets/46ef559d-9666-4307-be38-f40768e6c88d)
![call_for_evidence_outcome-3](https://github.com/user-attachments/assets/152adcb3-5bba-447d-a3b2-ae6b3da19d3e)

![Screenshot 2025-06-17 at 14 24 51](https://github.com/user-attachments/assets/f9206de7-2b97-4d16-ba58-51ea68c9bf2d)

deborahchua committed:
Update README.md and dev controller examples

Similar to consultations, call for evidence documents have different
stages:
- unopened call for evidence
- open call for evidence
- closed call for evidence
- call for evidence outcome

deborahchua committed:
Add route, basic controller and placeholder view

deborahchua committed:
Copy call_for_evidence view from government-frontend

The view has been directly copied over from government-frontend.
It will be updated to match the new data structure in a later
commit.

Commit audit trail:
- https://github.com/alphagov/government-frontend/blob/f1f866238d476a4281cd79fdbab2d671c6d46619/app/views/content_items/call_for_evidence.html.erb

deborahchua committed:
Fix lint errors in view

Ran `bundle exec erb_lint --lint-all --autocorrect`

deborahchua committed:
Update @content_item to content_item in view

deborahchua committed:
Call Metadata component directly in the view

This keeps it consistent with the approach in other document types,
which call the metadata component directly.

It has also been updated to use `extra_headers`, which ensures that
the open graph meta tags are added to the page source correctly.
We also need to add in the description tag as the machine_readable_metadata
component only adds the og tag i.e. <meta name="og:description" content="description">

Audit trail:
machine_readable_metadata_helper.rb: https://github.com/alphagov/government-frontend/blob/0adbc42e5591f9ad914791306b1ae2558f9d3cfc/app/helpers/machine_readable_metadata_helper.rb

deborahchua committed:
Update view to use heading component

This is rather than porting across the `heading_and_context` presenter
in government-frontend.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/0adbc42e5591f9ad914791306b1ae2558f9d3cfc/app/presenters/content_item/heading_and_context.rb

deborahchua committed:
Add initial CallForEvidence model

Attachments do not exist in Frontend yet and will be added in a
subsequent commits.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/0adbc42e5591f9ad914791306b1ae2558f9d3cfc/app/presenters/call_for_evidence_presenter.rb#L5-L8

deborahchua committed:
Add Attachments concern and tests

These methods are used in the `attachments_list` partial, which will
also be included on consultation and publication pages.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/app/presenters/content_item/attachments.rb
- https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/test/presenters/content_item/attachments_test.rb

deborahchua committed:
Update CallForEvidence model to include attachments

Audit trail:
- https://github.com/alphagov/government-frontend/blob/0adbc42e5591f9ad914791306b1ae2558f9d3cfc/app/presenters/call_for_evidence_presenter.rb#L2

deborahchua committed:
Update view to use publisher_metdata partial

This was introduced in https://github.com/alphagov/frontend/commit/ad511e05ad4d44aea1a7b6c2aec510647502605e

deborahchua committed:
Update view to use Withdrawable concern methods

These were introduced in https://github.com/alphagov/frontend/commit/4b1d2d4c4b2ba76b13775e26bbe2f6ee61fa3d8e
and can be used in the withdrawn notice banner.

deborahchua committed:
Add CallForEvidence presenter

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L11-L29
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/test/presenters/call_for_evidence_presenter_test.rb#L26-L61

deborahchua committed:
Update CallForEvidence controller to use presenter

deborahchua committed:
Update view to use presenter

deborahchua committed:
Update model to return document outcome details

There were no existing tests in government-frontend and have now
been added.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L47-L49

deborahchua committed:
Add attachments_list shared partial

It is also included on consultation and publication pages.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/c229e1c71ffb8424c1e072fa99db33cecf13e199/app/views/content_items/_attachments_list.html.erb

deborahchua committed:
Update model to return document dates

There were no existing tests and have now been added.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L11-L17

deborahchua committed:
Update model to return document types

Call for evidence documents have different stages (by document type).
They are used in the view to determine which sections of information
are shown.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L31-L45
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/test/presenters/call_for_evidence_presenter_test.rb#L63-L75

deborahchua committed:
Update model to handle document attachments

The method `attachments_with_details` has been renamed to
`attachments_with_details_count` to make it clearer what it is
returning.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L51-L65
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/test/presenters/call_for_evidence_presenter_test.rb#L26-L97

deborahchua committed:
Update model to check for external url

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L67-L73
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/test/presenters/call_for_evidence_presenter_test.rb#L99-L105

deborahchua committed:
Update model to handle document contact details

The methods `ways_to_respond?` and `response_form?` are called in
the view to determine whether certain sections are shown on the
page. They were originally returning the first contact details
available in their check, but we only care about whether they are
present or not.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/app/presenters/call_for_evidence_presenter.rb#L75-L97
- https://github.com/alphagov/government-frontend/blob/ef31d532d50a1ae80e8831f2f6eac60f54e06ec9/test/presenters/call_for_evidence_presenter_test.rb#L107-L148

deborahchua committed:
Update view with new location of locale keys

Frontend has a slightly different yaml structure and locale keys for
call for evidence content now sit under `formats`.

deborahchua committed:
Update how govspeak is called in document view

deborahchua committed:
Update view to use share links helper method

deborahchua committed:
Update view with attachments_list partial location

deborahchua committed:
Add request tests

deborahchua committed:
Copy integration test as system test

This is a straight copy from government-frontend, and will be updated
in the following commits.

Audit trail:
- https://github.com/alphagov/government-frontend/blob/cadb97a620b03c18037aec2c4df72546048e6a87/test/integration/call_for_evidence_test.rb

deborahchua committed:
Convert from minitest to rspec

All tests have also been commented out, and incrementally
introduced/updated in following commits.

deborahchua committed:
Add system tests for generic call for evidence features

These features apply to all types of call for evidence documents.

deborahchua committed:
Add system tests for open_call_for_evidence features

deborahchua committed:
Add system tests for unopened call for evidence features

deborahchua committed:
Add system tests for closed_call_for_evidence features

deborahchua committed:
Add system tests for call_for_evidence_outcome features

deborahchua committed:
Add locale keys for open_call_for_evidence schema

Ran rake "consolidation:copy_translation[content_item.schema_name.open_call_for_evidence, formats.open_call_for_evidence.name]"

Audit trail:
- https://github.com/alphagov/government-frontend/tree/cadb97a620b03c18037aec2c4df72546048e6a87/config/locales

deborahchua committed:
Add locale keys for closed_call_for_evidence schema

Ran rake "consolidation:copy_translation[content_item.schema_name.closed_call_for_evidence, formats.closed_call_for_evidence.name]"

Audit trail:
- https://github.com/alphagov/government-frontend/tree/cadb97a620b03c18037aec2c4df72546048e6a87/config/locales

deborahchua committed:
Add locale keys for call_for_evidence_outcome schema

Ran rake "consolidation:copy_translation[content_item.schema_name.call_for_evidence_outcome, formats.call_for_evidence_outcome.name]"

Audit trail:
- https://github.com/alphagov/government-frontend/tree/cadb97a620b03c18037aec2c4df72546048e6a87/config/locales

deborahchua committed:
Fix html passed into summary_banner component

The summary_banner isn't designed to accept additional text aready
wrapped in <p> tags - this creates nested <p> tags which is not
valid (and fails CI).
Adding line breaks should be a sufficient workaround.

Reference:
- https://github.com/alphagov/govuk_publishing_components/blob/main/app/views/govuk_publishing_components/components/_summary_banner.html.erb#L22

deborahchua committed:
Move logic out of view into presenter

The view renders different text depending on the document type
(unopened, open, closed, outcome). This is an attempt at making the
view less confusing, and taking care of some of the logic in the
presenter.

We will dry up the code in subsequent commits.

deborahchua committed:
Refactor summary logic in view

There is some additional logic to add a </section> tag if the document
type is call_for_evidence, which isn't clear to follow what exactly
it is closing. This has now been moved into the previous control
flow structure.

In doing so, this has introduced some repetition in calling the
summary_banner component. There also is a hefty amount of logic in
the capture methods. Moving this all into a partial should tidy things
up a bit (see next commit).

deborahchua committed:
Move summary into partial

KludgeKML commented on app/presenters/call_for_evidence_presenter.rb:
```diff
@@ -3,16 +3,30 @@ def opening_date
+    opening_date_midnight? ? I18n.t("formats.call_for_evidence.on") : I18n.t("formats.call_for_evidence.at")
+  end
+
+  def opens_close_or_ran
```
Nitpick: perhaps should be `opens_closes_or_ran` ?

KludgeKML replied:
Hmm, this is still showing the old names for me. Did it get squashed into that commit correctly?

deborahchua replied:
Ah sorry, I initially misread your suggestion! It's now updated in https://github.com/alphagov/frontend/pull/4855/commits/837ce9d9f7aef1d0ea159db5638ad42a32c83a5a and subsequent commits.

KludgeKML reviewed (changes requested):
A few small notes. You could also arguably roll the fixes to move the call_for_evidence pluralisation keys into the name: key into the commits where they're moved, rather than having one commit to bring them in and another to fix them up to match PR 4848. Otherwise looking very good, well laid out and some nice improvements!

leenagupte commented on app/models/call_for_evidence.rb:
```diff
@@ -0,0 +1,10 @@
+  include SinglePageNotificationButton
+  include Updatable
+
+  def contributors
```
Keith it's a bit of a mess, but it comes from the [`from` method ](https://github.com/alphagov/government-frontend/blob/eff89790a0a122b1fc944bff6641ed41e665efa1/app/presenters/content_item/linkable.rb#L4) which is rendered by the `publisher_metadata_with_logo` partial.

I can't find the commit, but I think that we decided to make it obvious which types of link should be included for each document_type, and this became the `contributors` method.

KludgeKML replied:
Ah, right. Thanks!

deborahchua committed:
Update method in attachments_list shared partial

deborahchua committed:
Add locale keys for call_for_evidence schema

Ran rake "consolidation:copy_translation[content_item.schema_name.call_for_evidence, formats.call_for_evidence.name]"

Audit trail:
- https://github.com/alphagov/government-frontend/tree/0adbc42e5591f9ad914791306b1ae2558f9d3cfc/config/locales

deborahchua committed:
Add locale keys for call for evidence content

Ran rake "consolidation:copy_translation[call_for_evidence, formats.call_for_evidence]"

Audit trail:
- https://github.com/alphagov/government-frontend/tree/cadb97a620b03c18037aec2c4df72546048e6a87/config/locales

deborahchua commented:
Thanks @KludgeKML , this is ready for a re-review.

> A few small notes. You could also arguably roll the fixes to move the call_for_evidence pluralisation keys into the name: key into the commits where they're moved, rather than having one commit to bring them in and another to fix them up to match PR 4848. Otherwise looking very good, well laid out and some nice improvements!

Good point thank you! It's now in the original commit [Add locale keys for call_for_evidence schema](https://github.com/alphagov/frontend/pull/4855/commits/ff9d6759b9f0feaa3d98d14169bb57fb541ab5c6) - there were conflicts to resolve in https://github.com/alphagov/frontend/pull/4855/commits/836b7e1c122f850738e905d8324da866e5d59c9b and https://github.com/alphagov/frontend/pull/4855/commits/1b84a2a9462959ad9a320e70d46d01ccee95863c so would appreciate another check on these please 🙏 

I've also updated commit [Add Attachments concern and tests](https://github.com/alphagov/frontend/pull/4855/commits/4ea791ce314c21b9a786d115efb8f06a4a9564c8) since your last review - I didn't realise the file changed recently so the code and audit trail has been updated to reflect the new changes.

KludgeKML reviewed (changes requested):
LGTM except that we should probably work out the print_link translation thing in the gem, because it really should be working there so we shouldn't add a workaround for it here.

deborahchua commented:
> LGTM except that we should probably work out the print_link translation thing in the gem, because it really should be working there so we shouldn't add a workaround for it here.

I've dropped the commit and rebased against `main` which now has the changes from https://github.com/alphagov/frontend/pull/4869

KludgeKML reviewed (approved):
LGTM!
